plugins/commands.js

```diff
@@ -0,0 +1,183 @@
+'use strict';
+
+const joi = require('joi');
+const boom = require('boom');
+const requestApi = require('request');
+const SCHEMA_COMMAND_NAMESPACE = joi.string().label('Command namespace');
+const SCHEMA_COMMAND_NAME = joi.string().label('Command name');
+const SCHEMA_COMMAND_VERSION = joi.string().label('Command version');
+const DEFAULT_TTL = 21 * 24 * 60 * 60 * 1000; // 3 weeks
+const DEFAULT_BYTES = 1024 * 1024 * 1024; // 1GB
+
+/**
+ * Determine if a pipeline can publish the command
+ * @method canPublishCommand
+ * @param  {String}  namespace  Command namespace
+ * @param  {String}  name       Command name
+ * @param  {Integer} pipelineId pipelineId
+ * @param  {String}  baseurl    API URL
+ * @return {Promise}
+ */
+function canPublishCommand(namespace, name, pipelineId, baseurl) {
+    return new Promise((resolve, reject) => {
+        requestApi({
+            uri: `${baseurl}/v4/commands`,
+            method: 'GET',
+            json: true
+        },
+        (err, response) => {
+            if (err) {
+                return reject(err);
+            }
+
+            return resolve(response);
+        });
+    }).then((response) => {
+        const commands = response.body.filter(element =>
+            element.namespace === namespace && element.name === name
+        );
+
+        // If commands.length === 0, it is first time to publish namespace/name command.
+        if (commands.length === 0 || commands[0].pipelineId === pipelineId) {
+            return Promise.resolve(true);
+        }
+
+        return Promise.resolve(false);
+    });
+}
+
+/**
+ * Commands Plugin
+ * @method register
+ * @param  {Hapi}     server                Hapi Server
+ * @param  {Object}   options               Configuration
+ * @param  {Integer}  options.expiresInSec  How long to keep it around
+ * @param  {Integer}  options.maxByteSize   Maximum Bytes to accept
+ * @param  {Function} next                  Function to call when done
+ */
+exports.register = (server, options, next) => {
+    const cache = server.cache({
+        segment: 'commands',
+        expiresIn: parseInt(options.expiresInSec, 10) || DEFAULT_TTL
+    });
+
+    server.expose('stats', cache.stats);
+    server.route([{
+        method: 'GET',
+        path: '/commands/{namespace}/{name}/{version}',
+        config: {
+            description: 'Get command binary',
+            notes: 'Get a script or binary of specific command',
+            tags: ['api', 'commands'],
+            plugins: {
```

Comment:
I think we want some `auth` for GET too. We just added that for logs and artifacts recently.

Reply:
I added `auth` like artifacts to GET.

Reply:
Only user can fetch it? Do we allow a build to fetch it?

Reply:
Yes, we want to use commands in a build, so we have to allow a build to fetch it. Fixed it.